modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,239 @@
+ *
+ * <p>Keeps up-to-date lists of primary replicas by partition for every table, avoiding expensive placement driver calls in most cases.
+ */
+public class ClientPrimaryReplicaTracker implements EventListener<EventParameters> {
+    private static final int LRU_CHECK_FREQ_MILLIS = 60 * 60 * 1000;
+
+    private final ConcurrentHashMap<Integer, Holder> primaryReplicas = new ConcurrentHashMap<>();
+
+    private final AtomicLong updateCount = new AtomicLong();
+
+    private final PlacementDriver placementDriver;
+
+    private final IgniteTablesInternal igniteTables;
+
+    private final HybridClock clock;
+
+    private final AtomicLong lruCheckTime = new AtomicLong(0);
+
+    private final EventProducer<CatalogEvent, CatalogEventParameters> catalogEventProducer;
+
+    /**
+     * Constructor.
+     *
+     * @param placementDriver Placement driver.
+     * @param igniteTables Ignite tables.
+     * @param clock Hybrid clock.
+     */
+    public ClientPrimaryReplicaTracker(
+            PlacementDriver placementDriver,
+            IgniteTablesInternal igniteTables,
+            EventProducer<CatalogEvent, CatalogEventParameters> catalogEventProducer,
+            HybridClock clock) {
+        this.placementDriver = placementDriver;
+        this.igniteTables = igniteTables;
+        this.catalogEventProducer = catalogEventProducer;
+        this.clock = clock;
+    }
+
+    /**
+     * Gets primary replicas by partition for the table.
+     *
+     * @param tableId Table ID.
+     * @return Primary replicas for the table, or null when not yet known.
+     */
+    public CompletableFuture<List<ReplicaHolder>> primaryReplicasAsync(int tableId) {
+        return primaryReplicas.compute(tableId, (id, hld) -> {
+            if (hld == null || hld.replicas.isCompletedExceptionally()) {
+                return new Holder(initReplicasForTableAsync(id));
+            }
+
+            hld.lastAccessTime = System.currentTimeMillis();
+            return hld;
+        }).replicas;
+    }
+
+    long updateCount() {
+        return updateCount.get();
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void start() {
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        catalogEventProducer.listen(CatalogEvent.TABLE_DROP, (EventListener) this);
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void stop() {
+        catalogEventProducer.removeListener(CatalogEvent.TABLE_DROP, (EventListener) this);
+        placementDriver.removeListener(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        primaryReplicas.clear();
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> initReplicasForTableAsync(Integer tableId) {
+        try {
+            // Initially, request all primary replicas for the table.
+            // Then keep them updated via PRIMARY_REPLICA_ELECTED events.
+            return igniteTables
+                    .tableAsync(tableId)
```

Comment:
Why do you need tableAsync()? TableId is available though catalog. Partitions is a zone attribute
`org.apache.ignite.internal.catalog.descriptors.CatalogZoneDescriptor#partitions` where corresponding zoneId is also available in org.apache.ignite.internal.catalog.descriptors.CatalogTableDescriptor#zoneId. Thus you probably need catalog but not table and especially InternalTable.

Reply:
Thank you, good point. I've added catalog later and did not notice that `partitions` is available from there.